krm-functions/lib/kptfile/v1/kpt.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2023 Nephio.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+	kptv1 "github.com/GoogleContainerTools/kpt/pkg/api/kptfile/v1"
+	"sigs.k8s.io/yaml"
+)
+
+type KptFile interface {
+	// Marshal serializes the value provided into a YAML document based on "sigs.k8s.io/yaml".
+	// The structure of the generated document will reflect the structure of the value itself.
+	Marshal() ([]byte, error)
+	// ParseKubeObject returns a fn sdk KubeObject; if something failed an error
+	// is returned
+	ParseKubeObject() (*fn.KubeObject, error)
+	// GetKptFile returns the Kptfile as a go struct
+	GetKptFile() *kptv1.KptFile
+	// SetConditions sets the conditions in the kptfile. It either updates the entry if it exists
+	// or appends the entry if it does not exist.
+	SetConditions(...kptv1.Condition)
+	// DeleteCondition deletes the condition equal to the conditionType if it exists
+	DeleteCondition(ct string)
+	// GetCondition returns the condition for the given ConditionType if it exists,
+	// otherwise returns nil
+	GetCondition(ct string) *kptv1.Condition
+	// GetConditions returns all the conditions in the kptfile. if not initialized it
+	// returns an emoty slice
+	GetConditions() []kptv1.Condition
+}
+
+// New creates a new parser interface
+// It expects a raw byte slice as input representing the serialized yaml file
+func New(b string) (KptFile, error) {
+	k := &kptv1.KptFile{}
+	if err := yaml.Unmarshal([]byte(b), k); err != nil {
+		return nil, err
+	}
+	return &kptFile{
+		kptfile: k,
+	}, nil
+}
+
+type kptFile struct {
+	m       sync.RWMutex
```

Comment:
I am not sure the mutex is going to do what you hope. When running in a function, we are typically single-threaded anyway.

Are you intending this library to *also* be used by the specializer controllers? That is the concurrency we need to protect against - multiple specializers tweaking conditions at the same time. Unless all specializers run in the same binary (and we use leader election for that binary), we can't protect from that in this library. Instead, that has to be protected by Porch.

Consider:
 - controller A executes, reads the package
 - controller B executes, reads the package
 - controller A does its thing, and updates the conditions 
 - controller B does its thing, and updates the conditions - oops, it just overwrote controller A's conditions

Generally K8s uses optimistic concurrency. So, when the update is executed, it includes the original object revision; the server kicks it back if the object has since been updated. See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#concurrency-control-and-consistency.

Ultimately, we need something like that. We may be able to get away with less in R1, though, but it will have race conditions.

Reply:
the goal was not to solve the concurrency issue we have, but rather having a way to protect if ever needed. I dot believe it harms us either, but I agree it should be solved by porch.

Reply:
Yeah. In fact in R1, the constraints I list above are probably true, so this will help in that case.